cpp/examples/arrow/exec_plan_examples.cc

```diff
@@ -0,0 +1,1122 @@
+
+    cp::ExecContext exec_context(arrow::default_memory_pool(),
+                                 ::arrow::internal::GetCpuThreadPool());
+
+    ARROW_ASSIGN_OR_RAISE(std::shared_ptr<cp::ExecPlan> plan,
+     cp::ExecPlan::Make(&exec_context));
+    arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> sink_gen;
+
+    ARROW_ASSIGN_OR_RAISE(
+        cp::ExecNode * source,
+        cp::MakeExecNode("source",
+            plan.get(), {},
+                cp::SourceNodeOptions{
+                    input.schema,
+                    input.gen(/*parallel=*/true)}));
+
+    cp::SelectKOptions options = cp::SelectKOptions::TopKDefault(/*k=*/2, {"i32"});
+
+    ARROW_ASSIGN_OR_RAISE(
+        cp::ExecNode * k_sink_node,
+        cp::MakeExecNode("select_k_sink",
+            plan.get(), {source},
+                cp::SelectKSinkNodeOptions{options, &sink_gen}));
+
+    k_sink_node->finished().Wait();
+
+    std::shared_ptr<arrow::RecordBatchReader> sink_reader = cp::MakeGeneratorReader(
+        arrow::schema({arrow::field("i32", arrow::int32()),
+                       arrow::field("str", arrow::utf8())}),
+        std::move(sink_gen),
+        exec_context.memory_pool());
+
+    // // // validate the ExecPlan
+    ABORT_ON_FAILURE(plan->Validate());
+    // // // start the ExecPlan
+    ABORT_ON_FAILURE(plan->StartProducing());
+
+    // // collect sink_reader into a Table
+    std::shared_ptr<arrow::Table> response_table;
+
+    ARROW_ASSIGN_OR_RAISE(response_table,
+    arrow::Table::FromRecordBatchReader(sink_reader.get()));
+
+    PRINT_LINE("Results : " << response_table->ToString());
+
+    // // plan stop proudcing
+    plan->StopProducing();
+    // // plan mark finished
+    plan->finished().Wait();
+
+    return arrow::Status::OK();
+}
+
+arrow::Status scan_filter_write_example() {
+    cp::ExecContext exec_context(arrow::default_memory_pool(),
+                                 ::arrow::internal::GetCpuThreadPool());
+
+    // ensure arrow::dataset node factories are in the registry
+    arrow::dataset::internal::Initialize();
+
+    ARROW_ASSIGN_OR_RAISE(std::shared_ptr<cp::ExecPlan> plan,
+     cp::ExecPlan::Make(&exec_context));
+
+    std::shared_ptr<arrow::dataset::Dataset> dataset = CreateDataset();
+
+    auto options = std::make_shared<arrow::dataset::ScanOptions>();
+    // sync scanning is not supported by ScanNode
+    options->use_async = true;
+    // specify the filter
+    //cp::Expression b_is_true = cp::field_ref("b");
+    //options->filter = b_is_true;
+    // empty projection
+    options->projection = Materialize({});
+
+    cp::ExecNode *scan;
+
+    auto scan_node_options = arrow::dataset::ScanNodeOptions{dataset, options};
+
+    ARROW_ASSIGN_OR_RAISE(scan, cp::MakeExecNode("scan", plan.get(), {},
+    scan_node_options));
+
+    arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> sink_gen;
+
+    std::string root_path = "";
+    std::string uri = "file:///Users/vibhatha/sandbox/test";
```

Comment:
Remember to un-hardcode this :) 

We can link to gflags for command line argument parsing if necessary.

Reply:
Removed hardcoded values and added cmd args.